nexus/src/app/external_endpoints.rs

```diff
@@ -595,6 +648,125 @@ impl rustls::server::ResolvesServerCert for NexusCertResolver {
+    /// requests to restrict access to a Silo only via one of its endpoints.
+    ///
+    /// Normally, this works as follows: a client (whether a browser, CLI, or
+    /// otherwise) would be given a Silo-specific DNS name to use to reach one
+    /// of our external endpoints.  They'd use our own external DNS servers
+    /// (mostly likely indirectly) to resolve this to one of Nexus's external
+    /// IPs.  Clients then put that DNS name in either the "host" header (in
+    /// HTTP 1.1) or the URL's authority section (in HTTP 2 and later).
+    ///
+    /// In development, we do not assume that DNS has been set up properly.
+    /// That means we might receive requests that appear targeted at an IP
+    /// address or maybe are missing these fields altogether.  To support that
+    /// case, we'll choose an arbitrary Silo.
+    pub fn endpoint_for_request(
+        &self,
+        rqctx: &dropshot::RequestContext<Arc<ServerContext>>,
+    ) -> Result<Arc<ExternalEndpoint>, Error> {
+        let log = &rqctx.log;
+        let rqinfo = &rqctx.request;
+        let requested_authority = self
+            .authority_for_request(rqinfo)
+            .map_err(|e| Error::invalid_request(&format!("{:#}", e)))?;
+
+        let requested_host = requested_authority.host();
+        let log = log.new(o!("server_name" => requested_host.to_string()));
+        trace!(&log, "determining endpoint");
+
+        // If we have not successfully loaded the endpoint configuration yet,
+        // there's nothing we can do here.  We could try to do better (e.g., use
+        // the recovery Silo?).  But if we failed to load endpoints, it's likely
+        // the database is down, and we're not going to get much further anyway.
+        let endpoint_config = self.background_tasks.external_endpoints.borrow();
+        let endpoints = endpoint_config.as_ref().ok_or_else(|| {
+            error!(&log, "received request with no endpoints loaded");
+            Error::unavail("endpoints not loaded")
+        })?;
+
+        // See if there's an endpoint for the requested name.  If so, use it.
+        if let Some(endpoint) = endpoints.by_dns_name.get(requested_host) {
+            trace!(
+                &log,
+                "received request for endpoint";
+                "silo_name" => ?endpoint.db_silo.name(),
+                "silo_id" => ?endpoint.silo_id,
+            );
+
+            return Ok(endpoint.clone());
+        }
+
+        // There was no endpoint for the requested name.  This should generally
+        // not happen in deployed systems where we expect people to have set up
+        // DNS to find the external endpoints.  But in development, we don't
+        // always have DNS set up.  People may use an IP address to get here.
+        // As a best-effort, we'll pick a default endpoint for this case.  (We
+        // may want to put this behind a policy knob that differs in
+        // production.)
```

Comment:
The obvious question is what is the worst-case scenario where this does comes up in a production system? Seems fine because you either have creds for that silo or you don't, and if you don't, you can't do anything. Might be good to say that?

Reply:
Updated this in 6991974e.  It's a little wishy-washy because it's describing what the current callers do in a case that should never happen.  i.e., it's not the impact of this code per se, and it's subject to change in the near future.  But agreed someone may want to better understand this.